### ISSUE TITLE
The vm does not enforce that `deferred` cannot be a type name

### ISSUE THREAD

**Issue by eernstg** [dart-lang/sdk#31267]
`deferred` is a built-in identifier, and it is therefore a compile-time error to declare a type whose name is `deferred`. Nevertheless, the vm compiles the following without errors:
```dart
class deferred { }

main() {
  new deferred();
}
```

**Comment by crelier** [comment]
This is a duplicate of #25732.
It can be fixed, but is there a good reason to make `deferred` a built-in identifier? It seems like the spec was very heavy handed in making it so. I do not think there would be any parsing ambiguity in recognizing `deferred` in the import statement only. It is actually what the vm parser is doing now.

**Comment by eernstg** [comment]
> .. duplicate of #25732.
Oops, I searched, but I did not find that one.

> .. reason .. [for being] built-in identifier?

I can see your point, 'deferred' could be treated like `hide`, `on`, `of`, `show`, and `sync`, being mentioned concretely in the grammar, and still not being built-in identifiers. I just tried it, and Dart.g is perfectly happy about this change (so I'd expect this change to be not-so-disruptive in other Dart parsers as well).

However, I'm not convinced that we obviously want to turn as many of these "pseudo keywords" as possible into "pseudo identifiers" (that is, _almost_ like every other identifier). The fact is that we only have a couple of tests using, say, `deferred` or `abstract` as a variable name, and nobody else would consider it a huge improvement to use those words for "non-keywordy purposes" all over the place.

I've created an issue for the language team to fight it out. ;-)

**Comment by a-siva** [comment]
@eernstg could you mark this bug as dependent on the issue you created for the language team.

**Comment by eernstg** [comment]
Dependent? I can choose the state blocked, but there seems to be no standardized way to indicate on what. So here's a natural language reference: Marked this issue as blocked on the above-mentioned dart-lang-evolation#168.

**Comment by lrhn** [comment]
In the spec `deferred` is a built-in identifier. That makes the current behavior a bug. We have no current plan to change that - it's not on our list of Dart 2 language changes. We can talk about it, but unless we actually decide to do anything, and I see no *compelling* reason to do so at this point, it's still just a bug.
I don't think you should *expect* anything to change. So, I wouldn't consider this blocked.

That said, it's a parsing bug that the front-end will likely handle for the VM soon enough, so there is no reason to spend VM developer effort on it.

**Comment by crelier** [comment]
Fine, I will fix it, then. My point was that this restriction is pointless and relaxing it is harmless.
I will fix related issue #25733 at the same time.

**Comment by lrhn** [comment]
> Fine, I will fix it, then. My point was that this restriction is pointless and relaxing it is harmless.

I agree that relaxing it is harmless, and not implementing the restriction will likely affect nothing but language tests. If we were talking about adding the restriction now, we probably wouldn't do so. Keeping the restriction is equally harmless, so removing it from the specification and other implementations is also not worth the effort.
